crypto/fift/HashMap.cpp

```diff
@@ -0,0 +1,368 @@
+/*
+    This file is part of TON Blockchain Library.
+
+    TON Blockchain Library is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation, either version 2 of the License, or
+    (at your option) any later version.
+
+    TON Blockchain Library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with TON Blockchain Library.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "HashMap.h"
+#include "td/utils/Random.h"
+
+namespace fift {
+using td::Ref;
+
+DictKey::DictKey(vm::StackEntry se) {
+  auto tp = tp_ = se.type();
+  switch (tp) {
+    case Type::t_int:
+      ref_ = se.as_int();
+      break;
+    case Type::t_atom:
+      ref_ = se.as_atom();
+      break;
+    case Type::t_string:
+      ref_ = se.as_string_ref();
+      break;
+    case Type::t_bytes:
+      ref_ = se.as_bytes_ref();
+      break;
+    default:
+      tp_ = Type::t_null;
+  }
+  compute_hash();
+}
+
+DictKey::operator vm::StackEntry() const& {
+  switch (tp_) {
+    case Type::t_int:
+      return value<td::CntInt256>();
+    case Type::t_atom:
+      return value<vm::Atom>();
+    case Type::t_string:
+    case Type::t_bytes:
+      return {value<td::Cnt<std::string>>(), tp_ == Type::t_bytes};
+    default:
+      return {};
```

Comment:
Keys of unsupported types are converted to null. `hmap!+` with an unsupported key will silently do nothing. It's better to throw an exception in this case.